crates/ra_assists/src/assists/add_missing_impl_members.rs

```diff
@@ -160,21 +177,73 @@ fn add_body(fn_def: ast::FnDef) -> ast::FnDef {
     }
 }
 
+// FIXME: It would probably be nicer if we could get this via HIR (i.e. get the
+// trait ref, and then go from the types in the substs back to the syntax)
+// FIXME: This should be a general utility (not even just for assists)
+fn get_syntactic_substs(impl_block: ast::ImplBlock) -> Option<Vec<ast::TypeRef>> {
```

Comment:
Yeah, in general this is something that needs design. I suspect that the right solution here is to represent `hir::Trait`
as  `(TraitId, Subst)`, and not just `(TraitId)`. Ie, we should make all hir nodes for generic things carry a substitution all the times. This I think is the design Kotlin uses -- they just don't have things without substs.

Reply:
Getting the substs on the type level didn't really seem like the big problem to me -- I was more worried about then getting back to the syntax level. We could of course use `HirDisplay`, but that often doesn't result in resolvable names. That's why I wanted to use the syntax of the type reference written by the user. 

Of course an alternative way would be to synthesize a path that resolves -- which is in fact something we need to do anyway for all other type references in the function definition, since they might not resolve in the current module either (I'm hoping to work on that once we've got visibility working). Still, using what the user wrote seemed more elegant to me. (Not going through the HIR also has the advantage of working if the type arguments don't currently resolve.)

Are you ok with merging this for now? It won't be the 'final' design either way, but I think it's at least an improvement :slightly_smiling_face:

Reply:
I didn’t get a chance to review this properly yet, but what I saw seems good, so d+ :)

In general, I am always ok with doing anything with any assists, as they are nicely isolated and even somewhat panic-proof :)

Reply:
>Still, using what the user wrote seemed more elegant to me.

This is elegant in a sense that we don't need to write a lot of code to properly synthesise paths. Long term, we probably should extend `HirDisplay` to also return a list of required imports.